File: app/src/main/java/com/hover/stax/countries/CountryAdapter.java
package com.hover.stax.countries;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.hover.stax.R;
import com.hover.stax.databinding.CountryItemBinding;
import com.yariksoffice.lingver.Lingver;

import java.util.Locale;

public class CountryAdapter extends ArrayAdapter<String> {
	private static final String TAG ="CountryAdapter";

	private String[] countryCodes;

	public CountryAdapter(@NonNull  String[] codes, @NonNull Context context) {
		super(context, 0, codes);
		this.countryCodes = codes;
	}

	@NonNull
	@Override
	public View getView(int position, @Nullable View view, @NonNull ViewGroup parent) {
		ViewHolder holder;

		if (view == null) {
			CountryItemBinding binding = CountryItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
			view = binding.getRoot();

			holder = new ViewHolder(binding);
			view.setTag(holder);
		} else {
			holder = (ViewHolder) view.getTag();
		}

		holder.countryText.setText(getCountryString(countryCodes[position]));

		return view;
	}

	private static class ViewHolder {
		TextView countryText;

		public ViewHolder(CountryItemBinding binding){
			countryText = binding.countryTextId;
		}
	}

	String getCountryString(String code) {
		if(code.equals(codeRepresentingAllCountries())) return getContext().getString(R.string.all_countries_with_emoji);
		return getContext().getString(R.string.country_with_emoji, countryCodeToEmoji(code), getFullCountryName(code));
	}

	private static String getFullCountryName(String code){
		Locale loc = new Locale(Lingver.getInstance().getLanguage(), code);
		return loc.getDisplayCountry();
	}

	private static String countryCodeToEmoji(String countryCode) {
		int firstLetter = Character.codePointAt(countryCode, 0) - 0x41 + 0x1F1E6;
		int secondLetter = Character.codePointAt(countryCode, 1) - 0x41 + 0x1F1E6;
		return new String(Character.toChars(firstLetter)) + new String(Character.toChars(secondLetter));
	}
	public static String codeRepresentingAllCountries(){
		return "00";
	}

	@Nullable
	@Override
	public String getItem(int position) {
		if (getCount() > 0) return countryCodes[position];
		else return null;
	}

	@Override
	public int getCount() {
		return countryCodes.length;
	}

	@Override
	public long getItemId(int position) { return position; }

	@Override
	public int getItemViewType(int position) {
		return position;
	}

	public interface SelectListener {
		void countrySelect(String countryCode);
	}
}

File: app/src/main/java/com/hover/stax/countries/CountryAdapter.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.countries

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.hover.stax.R
import com.hover.stax.databinding.CountryItemBinding
import com.yariksoffice.lingver.Lingver
import java.util.*

class CountryAdapter(val codes: Array<String>, context: Context) : ArrayAdapter<String>(context, 0) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val holder: ViewHolder
        var view = convertView

        if (view == null) {
            val binding = CountryItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            view = binding.root

            holder = ViewHolder(binding)
            view.tag = holder
        } else
            holder = view.tag as ViewHolder

        holder.setText(getCountryString(codes[position]))

        return view
    }

    fun getCountryString(code: String?): String = context.getString(R.string.country_with_emoji, countryCodeToEmoji(code), getFullCountryName(code))

    private fun getFullCountryName(code: String?): String {
        if (code.isNullOrEmpty() || code == CountryAdapter.CODE_ALL_COUNTRIES) {
            return context.getString(R.string.all_countries_text)
        }
        val locale = Locale(Lingver.getInstance().getLanguage(), code)
        return locale.displayCountry
    }

    private fun countryCodeToEmoji(countryCode: String?): String {
        if (countryCode.isNullOrEmpty() || countryCode == CountryAdapter.CODE_ALL_COUNTRIES) { return context.getString(R.string.all_countries_emoji) }
        val firstLetter = Character.codePointAt(countryCode.uppercase(), 0) - 0x41 + 0x1F1E6
        val secondLetter = Character.codePointAt(countryCode.uppercase(), 1) - 0x41 + 0x1F1E6
        return String(Character.toChars(firstLetter)) + String(Character.toChars(secondLetter))
    }

    override fun getItem(position: Int): String? = if (count > 0) codes[position] else null

    override fun getCount(): Int = codes.size

    class ViewHolder(val binding: CountryItemBinding) {

        private var countryText: TextView = binding.countryTextId

        fun setText(country: String) {
            countryText.text = country
        }
    }

    interface SelectListener {
        fun countrySelect(countryCode: String)
    }

    companion object {
        const val CODE_ALL_COUNTRIES: String = "00"
    }
}